### PR TITLE
Try: Allow space between on menu items.

### PR DESCRIPTION
**Pull request by jasmussen**
This PR tries to add `space-between` to the navigation menu as an option. I also updated the icons. It works fine. Before:

before

After:

Screenshot 2021-01-13 at 14 02 26

Screenshot 2021-01-13 at 14 02 34

Here are the icons:

new-icons

What's missing?

- Right now the justification does nothing, as far as I can tell, on the vertical version of the menu. This is unchanged from before. Curious on what to do there.
- The icons are stored in a separate icons file, we should move them to the icon package so everyone can use them. This will require a little bit of refactoring, looks like.
- The logic to set the active justification icon in the menu is broken for the new addition, I'm sure I missed something. It just shows an obsolete dashicon:
Screenshot 2021-01-13 at 14 02 41

Penny for your thoughts?

### CODE REVIEW

**Comment by MaggieCabrera** [comment]
This will solve https://github.com/WordPress/gutenberg/issues/27132

**Comment by shaunandrews** [comment]
image


When I choose the new "Space between items" option, the icon in the toolbar seems like the wrong one.

**Comment by shaunandrews** [comment]
> Right now the justification does nothing, as far as I can tell, on the vertical version of the menu. This is unchanged from before. Curious on what to do there.

At the very least, I think the justification setting should not display for the vertical variation. It would be nice to add a height option to the vertical variant, and then allow for vertical spacing options.

**Comment by jasmussen** [comment]
I updated the "menu" icon to no longer use this Dashicon:

Screenshot 2021-01-14 at 10 31 38

But to use one from the WordPress library:

Screenshot 2021-01-14 at 10 35 24

Note, that menu icon should not be used in the menu at all, it should be showing the "Space Between" icon. I'm still chewing on that one, as well as the other good overnight feedback.

**Comment by jasmussen** [comment]
I refactored the code to use icons from the icon library. That means the following icons are now available to use elsewhere:

- `justifyLeft`
- `justifyCenter`
- `justifyRight`
- `justifySpaceBetween`

One hiccup, and it explains why the space between icon does not show up as active when selected. The following code sets the active icon in the toolbar:

```
						icon={
							attributes.itemsJustification
								? navIcons[
										`justify${ upperFirst(
											attributes.itemsJustification
										) }`
								  ]
								: navIcons.justifyLeft
						}
```

What it does is basically compare the list of icons (`justifyLeft`, `justifyCenter`, `justifyRight`, `justifySpaceBetween`) with the value that is set (`left`, `center, `right`, `space-between`). 

It works for `left`, because it prefixes `justify`, then makes the first letter (`l`) uppercase, becoming `justifyLeft`

But it breaks for `space-between`, because it becomes `justifySpace-between`. 

I feel like this is needlessly complex, and there should be a far simpler way of matching the value to icons.

@draganescu I know you worked on this, if you have a moment, any thoughts on how to best fix this? Would this work?

```
							{
								icon: justifySpaceBetween,
								title: __( 'Space between items' ),
								isActive:
									'space-between' ===
									attributes.itemsJustification,
								onClick: handleItemsAlignment(
									'justifySpaceBetween'
								),
							},
```

I'll keep tinkering in the mean time, but would appreciate pointers!

**Comment by jasmussen** [comment]
@mkaz Once we land this one, it would be nice to refactor this alignment dropdown so it's easy to port to Social Links, for example, that would solve #17275. CC: @ntsekouras as I think I recall you thinking about the same.

**Comment by jasmussen** [comment]
Okay, I got it working:

Screenshot 2021-01-14 at 14 24 09

But it's an ugly hack that can almost certainly be made much prettier. Shield your eyes: https://github.com/WordPress/gutenberg/pull/28169/commits/e34ef36b37a1bf52998e0945ec14a109c59540f4#diff-610616a7f92cf5567753ff6e54b42e95ffd4b9607984b72c814f56279e87fff4R62

If anyone who's less tired than I am would like to help me refactor that to something nicer, I'd much appreciate it. In the mean time, this should be ready for reviews!

File: packages/block-library/src/navigation/edit.js
Comment (draganescu) [inline packages/block-library/src/navigation/edit.js:68]: This is fine, but, as a nitpick, I would do it when getting the icon name instead of here. I think it's easier to read there. So, leave this as: 

```[ `items-justified-${ attributes.itemsJustification }` ]: attributes.itemsJustification,```

and update the call to be the normal css prop later in the file.

File: packages/block-library/src/navigation/edit.js
Comment (draganescu) [inline packages/block-library/src/navigation/edit.js:173]: Continued from the [above](https://github.com/WordPress/gutenberg/pull/28169/files#diff-610616a7f92cf5567753ff6e54b42e95ffd4b9607984b72c814f56279e87fff4R68) suggestion:

```
isActive:
	'space-between' ===
	attributes.itemsJustification,
onClick: handleItemsAlignment(
	'space-between'
),
```

File: packages/block-library/src/navigation/edit.js
Comment (draganescu) [inline packages/block-library/src/navigation/edit.js:139]: [Continued](https://github.com/WordPress/gutenberg/pull/28169/files#diff-610616a7f92cf5567753ff6e54b42e95ffd4b9607984b72c814f56279e87fff4R68) suggestion:

```
icon={
	attributes.itemsJustification
		? navIcons[
				`justify${ upperFirst(
					attributes.itemsJustification
				).replace( '-b', 'B' ) }`
		 ]
		: navIcons.justifyLeft
}
```

File: packages/block-library/src/navigation/edit.js
Comment (draganescu) [inline packages/block-library/src/navigation/edit.js:46]: Perhaps since for now this does not work for vertical we could try to replace with:

```
hasItemJustificationControls = attributes.orientation === 'horizontal',
```

**Review by draganescu — APPROVED** [review]
I tested this and it works great. If we merge this I'd remove the toolbar item for the vertical variation and make a separate issue to see how to handle item spacing in that case. Left a nitpick about obtaining the className for `space-between` .

Cool work!

**Comment by ntsekouras** [comment]
Don't merge yet :) - I'm looking at it a bit right now

**Review by ntsekouras — COMMENTED** [review]
Thanks for working on this Joen! I chimed in and made `buttons` use these nice new icons you created, for consistency.

In general as we have already discussed, we should make in the future a generic component for this: https://github.com/WordPress/gutenberg/pull/28169#issuecomment-760151821

The only reason I'm not approving now is that I pushed code as well. I believe it's ready to 🚢  though.

**Comment by jasmussen** [comment]
Thank you Nik, I tested this and it works as intended:

Screenshot 2021-01-14 at 18 19 08

Screenshot 2021-01-14 at 18 19 12

I restarted the tests, let's merge when they pass and work on the followups 👌

**Review by mkaz — APPROVED** [review]
I left a comment about your "ugly hack" feel free to switch using lodash kebab case if you want, but I don't see it as necessary since it also adds a small dependency, either way is fine.

This looks good, and I'll review the other issues to see what is next to get this implemented for the social links. :+1:

**Pull request merged**
